### PR TITLE
feat: add trunk alias with default to main/master/root

### PR DESCRIPTION
**Pull request by rslabbert**
See #1415 for context.

### CODE REVIEW

File: cli/src/cli_util.rs [hunk 1]
```diff
@@ -1883,6 +1890,14 @@ fn load_revset_aliases(
             }
         }
     }
+
+    if !aliases_map.contains_symbol(ALIAS_TRUNK) {
+        aliases_map.insert(
+            ALIAS_TRUNK,
+            "latest((present(main) | present(master) | present(root)) & remote_branches())",
```
Comment (ilyagr) [inline cli/src/cli_util.rs:1897]: Have you considered the alternative of putting this as a revset alias in a TOML file? The user can then just redefine it if they want.

You could put this in `https://github.com/martinvonz/jj/blob/main/cli/src/config/misc.toml` or, perhaps better, create `https://github.com/martinvonz/jj/blob/main/cli/src/config/aliases.toml` for both aliases and revset aliases.

Comment (rslabbert) [reply cli/src/cli_util.rs]: Amazing, I didn't realise that was there, definitely a cleaner/better approach.

**Comment by yuja** [comment]
> 1. Should we keep the alias named `trunk`?

I think function syntax `trunk()` or `main()` is better. I don't think a branch named "trunk" is uncommon. The `root` symbol has the same problem, and I prefer not to expand a set of reserved symbols.

File: cli/src/config/aliases.toml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+[aliases]
+# Placeholder: added by user
+
+[revset-aliases]
+trunk = "latest((present(main) | present(master) | present(root)) & remote_branches())"
```
[the remark this answers is not in the document]
Comment (martinvonz) [reply cli/src/config/aliases.toml:5]: Should we do `latest(remote_branches(main) | remote_branches(master) | root)` instead? Otherwise `main`/`master` resolves the the local branch, which doesn't necessarily point to the same place as the remote branch. The disadvantage is that it can resolve to many commits if there are many remotes, but `latest()` should generally take care of that.

**Comment by martinvonz** [comment]
> The `root` symbol has the same problem, and I prefer not to expand a set of reserved symbols.

I wouldn't mind if we made it a function instead.

File: cli/src/config/aliases.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (yuja) [reply cli/src/config/aliases.toml:5]: Maybe it's time to add string matcher syntax like `remote_branches(literal:main)`? `remote_branches(main)` can also match "maint". I think `mine()` revset would have the same problem.

[the remark this answers is not in the document]
Comment (martinvonz) [reply cli/src/config/aliases.toml:5]: Good point. "maint" is quite likely in practice. I don't think we need to block this PR on it, but we probably should add that somewhat soon. Let me know if you disagree.

I thought about for the `mine` revset, but it doesn't seem like much of a problem in practice there. I'm sure it happens that an email address is a suffix of another email address, but that's probably extremely rare, at least in the same repo.

**Comment by rslabbert** [comment]
Summary of updates:
1. Changed to a function.
2. Use `exact:"..."` syntax.
3. Use `remote_branches(x)` directly instead of using `present(x) & remote_branches()` as it resolves the same way.
4. Added tests for expected branch names, overriding the alias, when no trunk branch can be found, and also for testing that only exact matches are detected.
5. Updated changelog and revsets.md doc.

I think an error message plus hint for when a branch can't be found should be handled by the CLI code in a future PR. Currently the alias resolves to an empty revset if none of the branches resolve, so we can't easily hook into a "not found" error message. Alternatively, the function can be moved into `BUILTIN_FUNCTION_MAP` in `lib/src/revset.rs`, then we can add nice error handling on top of it. Open to suggestions.

File: cli/src/config/aliases.toml [hunk 3]
```diff
@@ -0,0 +1,13 @@
+[aliases]
+# Placeholder: added by user
+
+[revset-aliases]
+'trunk()' = '''
+latest(
+  remote_branches(exact:"main", exact:"origin") | 
+  remote_branches(exact:"master", exact:"origin") | 
+  remote_branches(exact:"root", exact:"origin") |
```
Comment (martinvonz) [inline cli/src/config/aliases.toml:9]: "root" is not a branch name, it's a special symbol (and might become a function instead as we talked about somewhere). So use just `root` instead of the `remote_branches()` expression

File: docs/revsets.md [hunk 4]
```diff
@@ -157,6 +157,25 @@ For example:
...
 
+### Built-in Aliases
+
+The following aliases are built-in and might be relied on to exist. In every case,
+these functions are defined as aliases to allow you to overwrite them in specific
+situations.
+
+* `trunk()`: Resolves to the head commit for the trunk branch of the `origin`
+  remote. The branches `main`, `master`, `root`, and `trunk` are tried in order.
+  If none of the branches exist, it evaluates to nothing.
```
Comment (martinvonz) [inline docs/revsets.md:168]: `root` always exists, so you can strike this sentence when you've updated the alias to use that symbol instead

File: cli/src/config/aliases.toml (hunk 3 above)
Comment (rslabbert) [reply cli/src/config/aliases.toml:9]: I was intending for a branch named root here instead of the root symbol. Would it make sense for this to default to root? For example, if we did `jj rebase -d 'trunk()'`, I don't think we want to had it rebase onto the root commit?

Comment (martinvonz) [reply cli/src/config/aliases.toml:9]: We should make `rebase` a no-op when the destination is an ancestor, but other than that I think `root` is useful as a fallback. In particular so we can default the log revset to `trunk()..` (roughly) and have it show all commits in repos without a `main` or `master` branch. I think we can also change the default argument for `jj git push` be `-r 'trunk()..@'`, which I think should be safe even if we use the root commit as fallback.

Comment (rslabbert) [reply cli/src/config/aliases.toml]: Thinking about the rebase case, we'll need to delineate between the local trunk and the remote trunk in order to ensure `trunk()..@` works. Do we need `trunk()` and `remote_trunk()` functions or should we maybe forgo `latest()` in the function so we have a full history to use?

Comment (rslabbert) [reply cli/src/config/aliases.toml]: The trunk alias added in this PR should definitely cover points 2-4 in the short/medium term.

* I currently define `jj log` in terms of `(trunk()..@) | (trunk()..@)-` under a `jj sl` alias. There might be a better default but it will for a stack show me the entire stack plus the root. For example for this PR:
```
@  ulzolrwp ruben.slabbert 8 hours ago push-zqwpmmvnmxyx f341b78b
│  feat: add trunk alias with default to main/master/root
◉  sptkkttk martinvonz 18 hours ago HEAD@git 5f3df4aa
│  revset: resolve "@" symbol's workspace id earlier (while parsing)
~
```
* Using `remote_branches()..@` doesn't seem to work because as soon as the branch has been submit once, it can't be pushed again with that alias since it's now a remote branch as well. So it might be suitable for an initial `jj git push --change` but not for subsequent updates (which is why I have submit `git push --change @` and push `git push -r "(trunk..@):: & branches(push)"` aliases defined separately). Additionally, I've been careful about using far reaching revsets like `remote_branches()` for default pushing since I don't want to accidentally force push an unintended branch. In terms of this PR, `trunk()..@` could work well. Or potentially `trunk()..@ & mine()`. There's also a specific case here that needs to be handled where the stack is a combination of single commit and multi-commit branches that I'd need to test revsets for.
* This alias is intended to serve as the configurable rebase destination and contains the heuristic. My `jj sync` alias is `rebase -d trunk()`. I'll add some context on the commit, and I suspect adding the default changes on top of this commit would make it more clear. I've held off on making the default changes though since I suspect that would require a bit off discussion on edge cases an implications, but I'm thinking it might be worth doing to make it easier to discuss the whole change set holistically.

In terms of `trunk()` vs a set of `immutable()` commits, I suspect an `immutable()` alias would also require some plumbing throughout the rewrite commands to ensure they're not changed to match user expectation. Additionally, in the context of defaults, a user might want to treat different types of branches as immutable/protected, not just the trunk branch (e.g. release branch, or deployment branches). `trunk()` here hits the sweet spot of being immediately usable and obvious as a default in my opinion but I'm not strongly attached to that.

**Comment by yuja** [comment]
> Alternatively, the function can be moved into `BUILTIN_FUNCTION_MAP` in `lib/src/revset.rs`, then we can add nice error handling on top of it.

I think `trunk()` will be a real function at some point (in order to read the configured default branch and remote for example), but using alias should be a good start. We can change the implementation when needed.

File: cli/src/config/aliases.toml (hunk 3 above)
Comment (martinvonz) [reply cli/src/config/aliases.toml]: @rslabbert, I sent #2135 for changing the default branches to push. Want to review that?

File: CHANGELOG.md [hunk 5]
```diff
@@ -44,6 +44,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   function parameter. For example, `author(foo@)` is now an error, and
   a revset alias `'revset-aliases.foo@' = '@'` will be ignored. 
 
+* A default revset-alias function `trunk()` now exists. If you previously defined
+  your own `trunk()` alias it will continue to overwrite the built-in one, but future
+  functionality will assume that the function resolves correctly.
```
Comment (ilyagr) [inline CHANGELOG.md:49]: I'd move everything after "...continue to overwrite the built-in one" to config.md or revsets.md. See also my comment on revsets.md.

File: docs/revsets.md [hunk 6]
```diff
@@ -157,6 +157,25 @@ For example:
 'user(x)' = 'author(x) | committer(x)'
 ```
 
+### Built-in Aliases
+
+The following aliases are built-in and might be relied on to exist. In every case,
+these functions are defined as aliases to allow you to overwrite them in specific
+situations.
```
Comment (ilyagr) [inline docs/revsets.md:164]: You can put "See [aliases.toml](./cli/src/config/aliases.toml) for the exact definitions" here.

File: docs/revsets.md [hunk 7]
```diff
@@ -157,6 +157,25 @@ For example:
...
+
+* `trunk()`: Resolves to the head commit for the trunk branch of the `origin`
+  remote. The branches `main`, `master`, and `trunk` are tried in order.
+  If none of the branches exist, it evaluates to the `root` symbol.
+
+  Can be overwritten per repository to point to the correct trunk branch if none
```
Comment (ilyagr) [inline docs/revsets.md:170]: I'd reword this:

You can override this per-repository as appropriate. If you do, make sure it always resolves to *some* commit, as `jj` may rely on this property. For example:

(your example)

**Review by ilyagr — COMMENTED** [review]
Thank you for working on this!

So far, I've had a few minor comments about the docs.

**Comment by ilyagr** [comment]
Is the current plan for `trunk()` to resolve to one commit? I know it's supposed to resolve to *at least* one, but what happens if there are two?

I wrote a related thought in https://github.com/martinvonz/jj/pull/2247#issuecomment-1714830932, assuming we only want one `trunk()` commit, though it might duplicate things this discussion already covers (I do need to reread it).

**Comment by rslabbert** [comment]
@ilyagr yes, `trunk()` should resolve to only a single commit, otherwise I feel like commands using it might become confusing. I believe the use of `|` and `latest` should handle that and I've added the `test_builtin_alias_trunk_matches_exactly_one_commit` test to confirm that in the presence of both `main` and `master`, only `main` is resolved.

**Review by martinvonz — APPROVED** [review]

File: CHANGELOG.md [hunk 8]
```diff
@@ -9,6 +9,11 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Breaking changes
 
+* A default revset-alias function `trunk()` now exists. If you previously defined
+  your own `trunk()` alias it will continue to overwrite the built-in one.
+  Check [revsets.toml](./cli/src/config/revsets.toml) and [revsets.md](./docs/revset.md)
```
[the remark this answers is not in the document]
Comment (rslabbert) [reply CHANGELOG.md]: I tend to add is as a habit so it immediately disambiguates between local links and external links. That being said in most cases like this isn't not an issue so happy to remove it for consistency.

**Review by yuja — APPROVED** [review]
> `trunk()` should resolve to only a single commit, otherwise I feel like commands using it might become confusing. I believe the use of `|` and `latest` should handle that

Perhaps, it can be enforced by making `trunk()` a real function. We'll need to implement `exactly(x, count)` first. For now, I think a predefined alias is simple and good.

**Comment by rslabbert** [comment]
> Perhaps, it can be enforced by making `trunk()` a real function. We'll need to implement `exactly(x, count)` first. For now, I think a predefined alias is simple and good.

I'll make a note to get back around to that once we have the function plus some usage of this.
